File: android/app/src/main/kotlin/dev/heywood8/claudebuddy/Handshake.kt
package dev.heywood8.claudebuddy

import kotlinx.serialization.Serializable
import java.net.URLDecoder
import java.util.Base64

/** Plaintext, carries no secrets: an opaque host id and a fresh salt. */
@Serializable
data class Hello(val t: String = "hello", val v: Int = 1, val host: String, val hs: String)

/** We answer only if we know the host and no other host is active. */
@Serializable
data class Challenge(val t: String = "challenge", val v: Int = 1, val ps: String)

/**
 * First encrypted frame from the bridge. Opening it is our proof that the bridge holds the
 * key; there is no separate challenge-response step.
 */
@Serializable
data class Auth(val t: String = "auth", val name: String, val time: List<Long>)

/** First encrypted frame from us, and the mirror of the same proof. */
@Serializable
data class Ready(val t: String = "ready", val device: String, val proto: Int = 1)

@Serializable
data class Bye(val t: String = "bye", val reason: String)

object ByeReason {
    const val UNKNOWN_HOST = "unknown_host"
    const val BUSY = "busy"
    const val VERSION = "version"
    const val BAD_FRAME = "bad_frame"
    const val BAD_COUNTER = "bad_counter"
    const val SHUTDOWN = "shutdown"
}

/**
 * A paired bridge. Several may be stored; only one is served at a time.
 *
 * [hostId] is opaque rather than a machine name because it travels in the plaintext hello.
 */
data class PairedHost(val hostId: String, val key: ByteArray, val name: String) {
    override fun equals(other: Any?): Boolean =
        other is PairedHost && other.hostId == hostId

    override fun hashCode(): Int = hostId.hashCode()
}

/**
 * Parses the `cmb://pair?...` payload the bridge renders as a QR code.
 *
 * Deliberately hand-rolled rather than using `android.net.Uri`, so the same code runs under
 * plain JVM unit tests against the shared protocol vectors.
 */
object PairingCode {
    const val SCHEME = "cmb://pair"

    fun parse(text: String): PairedHost? {
        if (!text.startsWith("$SCHEME?")) return null
        val query = text.substring(SCHEME.length + 1)
        val values = mutableMapOf<String, String>()
        for (pair in query.split("&")) {
            val eq = pair.indexOf('=')
            if (eq <= 0) continue
            val name = pair.substring(0, eq)
            val value = runCatching {
                URLDecoder.decode(pair.substring(eq + 1), "UTF-8")
            }.getOrNull() ?: continue
            values.putIfAbsent(name, value)
        }

        val hostId = values["h"] ?: return null
        if (hostId.length != 32 || !hostId.all { it in '0'..'9' || it in 'a'..'f' }) return null
        val key = values["k"]?.let(::decodeBase64Url) ?: return null
        if (key.size != 32) return null
        return PairedHost(hostId = hostId, key = key, name = values["n"].orEmpty())
    }

    private fun decodeBase64Url(text: String): ByteArray? = runCatching {
        Base64.getUrlDecoder().decode(text.trimEnd('='))
    }.getOrNull()
}

File: bridge/Sources/cmbridge/Handshake.swift
import Foundation

/// Plaintext, carries no secrets: an opaque host id and a fresh salt.
struct Hello: Codable, Equatable {
    var t: String = "hello"
    var v: Int = 1
    let host: String
    let hs: String
}

/// The phone answers only if it knows the host and no other host is active.
struct Challenge: Codable, Equatable {
    var t: String = "challenge"
    var v: Int = 1
    let ps: String
}

/// First encrypted frame from the bridge. Opening it is the phone's proof that the bridge
/// holds the key; there is no separate challenge-response step.
struct Auth: Codable, Equatable {
    var t: String = "auth"
    let name: String
    /// Unix seconds and the host's UTC offset in seconds, so the phone can render the
    /// host's timestamps in the host's own clock.
    let time: [Int]
}

/// First encrypted frame from the phone, and the mirror of the same proof.
struct Ready: Codable, Equatable {
    var t: String = "ready"
    let device: String
    var proto: Int = 1
}

/// Everything a pairing QR code carries.
///
/// The bridge generates all of it and renders it; the phone only reads. A short typed code
/// was rejected during design: six digits is twenty bits, and a sniffed session would let
/// anyone brute-force the key offline in minutes.
struct PairingCode: Equatable {
    /// 32 lowercase hex characters. Opaque on purpose — it travels in the plaintext `hello`,
    /// so it must not leak a machine name.
    let hostID: String
    /// 32 bytes.
    let key: Data
    let hostName: String

    static let scheme = "cmb://pair"

    var url: String {
        var name = URLComponents()
        name.queryItems = [URLQueryItem(name: "n", value: hostName)]
        let encodedName = name.percentEncodedQuery ?? "n="
        return "\(Self.scheme)?h=\(hostID)&k=\(Base64URL.encode(key))&\(encodedName)"
    }

    static func generate(hostName: String) -> PairingCode {
        PairingCode(
            hostID: randomHex(16),
            key: randomBytes(32),
            hostName: hostName)
    }

    static func parse(_ text: String) -> PairingCode? {
        guard text.hasPrefix(scheme),
              let components = URLComponents(string: text),
              let items = components.queryItems else { return nil }
        let values = Dictionary(items.compactMap { item in
            item.value.map { (item.name, $0) }
        }, uniquingKeysWith: { first, _ in first })

        guard let hostID = values["h"],
              hostID.count == 32,
              hostID.allSatisfy({ $0.isHexDigit && !$0.isUppercase }),
              let keyText = values["k"],
              let key = Base64URL.decode(keyText),
              key.count == 32 else { return nil }
        return PairingCode(hostID: hostID, key: key, hostName: values["n"] ?? "")
    }

    private static func randomBytes(_ count: Int) -> Data {
        Data((0..<count).map { _ in UInt8.random(in: 0...255) })
    }

    private static func randomHex(_ byteCount: Int) -> String {
        randomBytes(byteCount).map { String(format: "%02x", $0) }.joined()
    }
}

enum Base64URL {
    static func encode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    static func decode(_ text: String) -> Data? {
        var padded = text
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        while padded.count % 4 != 0 { padded.append("=") }
        return Data(base64Encoded: padded)
    }
}
